// mcporter/dist/paths.js
import os from 'node:os';
import path from 'node:path';
import { expandHome } from './env.js';
const XDG_HOME_ENV = {
    config: 'XDG_CONFIG_HOME',
    data: 'XDG_DATA_HOME',
    state: 'XDG_STATE_HOME',
    cache: 'XDG_CACHE_HOME',
};
export function legacyMcporterDir() {
    return path.join(os.homedir(), '.mcporter');
}
export function mcporterDir(kind) {
    const raw = process.env[XDG_HOME_ENV[kind]];
    if (raw && raw.trim().length > 0) {
        const resolved = expandHome(raw.trim());
        if (path.isAbsolute(resolved)) {
            return path.join(resolved, 'mcporter');
        }
    }
    return legacyMcporterDir();
}
export function mcporterConfigCandidates() {
    const base = mcporterDir('config');
    const candidates = [path.join(base, 'mcporter.json'), path.join(base, 'mcporter.jsonc')];
    if (base !== legacyMcporterDir()) {
        const legacy = legacyMcporterDir();
        candidates.push(path.join(legacy, 'mcporter.json'), path.join(legacy, 'mcporter.jsonc'));
    }
    return candidates;
}
