// mcporter/dist/serve.js
import http from 'node:http';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { StreamableHTTPServerTransport } from '@modelcontextprotocol/sdk/server/streamableHttp.js';
import { CallToolRequestSchema, ErrorCode, ListToolsRequestSchema, McpError, } from '@modelcontextprotocol/sdk/types.js';
import { isKeepAliveServer } from './lifecycle.js';
import { MCPORTER_VERSION } from './version.js';
const TOOL_SEPARATOR = '__';
const DEFAULT_OBJECT_SCHEMA = { type: 'object' };
export const DEFAULT_SERVE_HTTP_HOST = '127.0.0.1';
export async function serveStdio(options) {
    const server = createBridgeServer(options);
    const transport = new StdioServerTransport();
    const closed = new Promise((resolve, reject) => {
        transport.onclose = () => resolve();
        transport.onerror = (error) => reject(error);
    });
    await server.connect(transport);
    await closed;
}
export async function serveHttp(options) {
    const httpServer = http.createServer((request, response) => {
        const url = new URL(request.url ?? '/', `http://${DEFAULT_SERVE_HTTP_HOST}`);
        if (url.pathname !== '/mcp') {
            response.writeHead(404).end('Not found');
            return;
        }
        const bridgeServer = createBridgeServer(options);
        const transport = new StreamableHTTPServerTransport({
            sessionIdGenerator: undefined,
        });
        response.on('close', () => {
            void transport.close().catch(() => { });
            void bridgeServer.close().catch(() => { });
        });
        void (async () => {
            await bridgeServer.connect(transport);
            await transport.handleRequest(request, response);
        })().catch((error) => {
            const message = error instanceof Error ? error.message : String(error);
            if (!response.headersSent) {
                response.writeHead(500, { 'content-type': 'text/plain; charset=utf-8' });
            }
            response.end(message);
        });
    });
    await new Promise((resolve, reject) => {
        httpServer.once('error', reject);
        httpServer.listen(options.port, options.host ?? DEFAULT_SERVE_HTTP_HOST, () => {
            httpServer.off('error', reject);
            resolve();
        });
    });
    return httpServer;
}
export function createBridgeServer(options) {
    const servedServers = selectServedServers(options.definitions, options.servers);
    if (servedServers.length === 0) {
        throw new Error('No keep-alive MCP servers are available to serve.');
    }
    const server = new McpServer({ name: 'mcporter-serve', version: MCPORTER_VERSION }, {
        capabilities: {
            tools: {},
        },
        instructions: 'MCPorter bridge exposing daemon-managed MCP servers. Tool names are namespaced as server__tool.',
    });
    server.server.setRequestHandler(ListToolsRequestSchema, async () => {
        const tools = [];
        for (const served of servedServers) {
            const listed = (await options.runtime.listTools(served.name, {
                includeSchema: true,
                autoAuthorize: true,
            }));
            for (const tool of listed) {
                tools.push({
                    name: encodeToolName(served.name, tool.name),
                    description: describeTool(served.name, tool.description),
                    inputSchema: normalizeInputSchema(tool.inputSchema),
                    outputSchema: normalizeOutputSchema(tool.outputSchema),
                });
            }
        }
        return { tools };
    });
    server.server.setRequestHandler(CallToolRequestSchema, async (request) => {
        const target = decodeToolName(request.params.name, servedServers);
        if (!target) {
            throw new McpError(ErrorCode.InvalidParams, `Unknown bridged tool '${request.params.name}'.`);
        }
        const result = await options.runtime.callTool(target.server, target.tool, {
            args: request.params.arguments,
        });
        return result;
    });
    return server;
}
export function selectServedServers(definitions, requested) {
    const keepAlive = definitions.filter(isKeepAliveServer);
    if (!requested || requested.length === 0) {
        return keepAlive.map((definition) => ({ name: definition.name, definition }));
    }
    const byName = new Map(keepAlive.map((definition) => [definition.name, definition]));
    return requested.map((name) => {
        const definition = byName.get(name);
        if (!definition) {
            throw new Error(`Server '${name}' is not configured for keep-alive and cannot be served by the daemon bridge.`);
        }
        return { name, definition };
    });
}
export function encodeToolName(server, tool) {
    return `${encodeToolNamePart(server, { escapeTrailingUnderscore: true })}${TOOL_SEPARATOR}${encodeToolNamePart(tool, { escapeLeadingUnderscore: true })}`;
}
export function decodeToolName(name, servedServers) {
    const candidates = servedServers
        .map((served) => ({
        name: served.name,
        prefix: `${encodeToolNamePart(served.name, { escapeTrailingUnderscore: true })}${TOOL_SEPARATOR}`,
    }))
        .toSorted((first, second) => second.prefix.length - first.prefix.length);
    for (const candidate of candidates) {
        if (!name.startsWith(candidate.prefix)) {
            continue;
        }
        const tool = decodeToolNamePart(name.slice(candidate.prefix.length));
        if (tool.length === 0) {
            return undefined;
        }
        return { server: candidate.name, tool };
    }
    return undefined;
}
function encodeToolNamePart(value, options = {}) {
    let encoded = value.replaceAll('%', '%25').replaceAll(TOOL_SEPARATOR, '%5F%5F');
    if (options.escapeLeadingUnderscore && encoded.startsWith('_')) {
        encoded = `%5F${encoded.slice(1)}`;
    }
    if (options.escapeTrailingUnderscore && encoded.endsWith('_')) {
        encoded = `${encoded.slice(0, -1)}%5F`;
    }
    return encoded;
}
function decodeToolNamePart(value) {
    return value.replaceAll('%5F', '_').replaceAll('%25', '%');
}
function describeTool(server, description) {
    if (!description) {
        return `Tool from MCPorter server '${server}'.`;
    }
    return `[${server}] ${description}`;
}
function normalizeInputSchema(schema) {
    if (isObjectSchema(schema)) {
        return schema;
    }
    return DEFAULT_OBJECT_SCHEMA;
}
function normalizeOutputSchema(schema) {
    if (isObjectSchema(schema)) {
        return schema;
    }
    return undefined;
}
function isObjectSchema(schema) {
    if (!schema || typeof schema !== 'object') {
        return false;
    }
    return schema.type === 'object';
}
