// mcporter/dist/daemon/request-utils.js
import { keepAliveIdleTimeout } from '../lifecycle.js';
export function ensureManaged(server, managedServers) {
    if (!managedServers.has(server)) {
        throw new Error(`Server '${server}' is not managed by the daemon.`);
    }
}
export function markActivity(server, activity) {
    const entry = activity.get(server);
    if (entry) {
        entry.connected = true;
        entry.lastUsedAt = Date.now();
    }
    else {
        activity.set(server, { connected: true, lastUsedAt: Date.now() });
    }
}
export async function evictIdleServers(runtime, managedServers, activity) {
    const now = Date.now();
    await Promise.all(Array.from(managedServers.entries()).map(async ([name, definition]) => {
        const timeout = keepAliveIdleTimeout(definition);
        if (!timeout) {
            return;
        }
        const entry = activity.get(name);
        if (!entry?.lastUsedAt) {
            return;
        }
        if (now - entry.lastUsedAt < timeout) {
            return;
        }
        await runtime.close(name).catch(() => { });
        activity.set(name, { connected: false });
    }));
}
export function shouldShutdownDaemonForIdle(lastActivityAt, now, idleTimeoutMs, activeRequests = 0) {
    return (activeRequests <= 0 &&
        typeof idleTimeoutMs === 'number' &&
        idleTimeoutMs > 0 &&
        now - lastActivityAt >= idleTimeoutMs);
}
export function daemonIdleWatcherInterval(idleTimeoutMs) {
    if (!idleTimeoutMs) {
        return 30_000;
    }
    return Math.min(30_000, Math.max(100, Math.floor(idleTimeoutMs / 2)));
}
export function buildErrorResponse(id, code, error) {
    let message = code;
    if (error instanceof Error) {
        message = error.message;
    }
    else if (typeof error === 'string') {
        message = error;
    }
    return {
        id,
        ok: false,
        error: {
            code,
            message,
        },
    };
}
