// clawhub/dist/deviceAuth.js
/**
 * GitHub Device Flow authentication for headless environments.
 *
 * Implements RFC 8628 / GitHub's Device Flow:
 * https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#device-flow
 *
 * This allows CLI authentication without a browser redirect to localhost,
 * enabling headless agents and remote servers to authenticate.
 */
const DEFAULT_SCOPE = "read write";
/**
 * Request a device code from the ClawHub device flow endpoint.
 */
export async function requestDeviceCode(config) {
    const url = new URL("/api/cli/device/code", config.apiUrl);
    const body = {
        scope: config.scope ?? DEFAULT_SCOPE,
        site_url: config.siteUrl,
    };
    if (config.clientId) {
        body.client_id = config.clientId;
    }
    const response = await fetch(url.toString(), {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            Accept: "application/json",
        },
        body: JSON.stringify(body),
    });
    if (!response.ok) {
        const text = await response.text().catch(() => "");
        throw new Error(`Device code request failed (${response.status}): ${text || response.statusText}`);
    }
    const data = (await response.json());
    if (!data.device_code || !data.user_code || !data.verification_uri) {
        throw new Error("Invalid device code response from server");
    }
    return data;
}
/**
 * Poll for the device flow token until the user completes authorization,
 * the code expires, or an unrecoverable error occurs.
 */
export async function pollForDeviceToken(config, deviceCode, options) {
    const url = new URL("/api/cli/device/token", config.apiUrl);
    const deadline = Date.now() + options.expiresIn * 1000;
    let interval = options.interval * 1000;
    const body = {
        device_code: deviceCode,
        grant_type: "urn:ietf:params:oauth:grant-type:device_code",
    };
    if (config.clientId) {
        body.client_id = config.clientId;
    }
    while (Date.now() < deadline) {
        await sleep(interval);
        const response = await fetch(url.toString(), {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Accept: "application/json",
            },
            body: JSON.stringify(body),
        });
        // Parse JSON once to avoid "body already read" errors
        const data = (await response.json().catch(() => ({})));
        if (response.ok && "access_token" in data && data.access_token) {
            return data;
        }
        const errorData = data;
        switch (errorData.error) {
            case "authorization_pending":
                // User hasn't completed auth yet — keep polling
                break;
            case "slow_down":
                // Server requests longer interval
                interval = (errorData.interval ?? Math.ceil(interval / 1000) + 5) * 1000;
                break;
            case "expired_token":
                throw new Error("Device code expired. Please try again.");
            case "access_denied":
                throw new Error("Authorization denied by user.");
            default:
                throw new Error(`Device flow error: ${errorData.error_description || errorData.error || "unknown error"}`);
        }
    }
    throw new Error("Device code expired (timeout). Please try again.");
}
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
